File: src/main/java/io/github/eingruenesbeb/yolo/SpicordManager.java
/*
 * This is program is a plugin for Minecraft Servers called "Yolo".
 * Copyright (C) 2023  eingruenesbeb
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 *
 * You can reach the original author via e-Mail: agreenbeb@gmail.com
 */

package io.github.eingruenesbeb.yolo;

import net.dv8tion.jda.api.requests.restaction.MessageCreateAction;
import net.dv8tion.jda.api.utils.messages.MessageCreateData;
import org.bukkit.entity.Player;
import org.bukkit.event.entity.PlayerDeathEvent;
import org.jetbrains.annotations.NotNull;
import org.spicord.SpicordLoader;
import org.spicord.api.addon.SimpleAddon;
import org.spicord.bot.DiscordBot;
import org.spicord.embed.EmbedParser;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Objects;

/**
 * This class is instantiated, when Spicord is also loaded as a Plugin. It manages every aspect regarding functionality
 * for sending Discord messages.
 */
public class SpicordManager {
    private final Yolo yolo = Yolo.getPlugin(Yolo.class);
    private boolean spicordBotAvailable;
    private DiscordBot spicordBot;
    private String messageChannelId;
    private String deathMessageTemplate;

    /**
     * Accessor for {@link SpicordManager#spicordBotAvailable}
     * @return If the bot is available for sending a message.
     */
    public boolean isSpicordBotAvailable() {
        return spicordBotAvailable;
    }

    /**
     * Accessor for {@link SpicordManager#spicordBot}
     *
     * @return (Spicord version of) The bot, that is used for sending the message, if provided.
     * @see DiscordBot
     */
    public DiscordBot getSpicordBot() {
        return spicordBot;
    }

    private void updateDeathMessageTemplate() throws IOException {
        try {
            deathMessageTemplate = Files.readString(Path.of(yolo.getDataFolder().getPath() + "/death_message.json"));
        } catch (IOException e) {
            // Shouldn't be happening, unless the file was deleted, after the plugin was loaded.
            // Default to the embedded message-config.
            // NPE shouldn't occur on embedded resources.
            deathMessageTemplate = new String(Objects.requireNonNull(yolo.getResource("death_message.json")).readAllBytes());
        }
    }

    /**
     * Tries to send a message to Discord created from the configured template, using Spicord and JDA.
     * @param player The player. This is used, to replace the variable {@code %player_name%} in the template. This is
     *               currently used in the event-listener.
     * @see YoloEventListener#onPlayerDeath(PlayerDeathEvent)
     */
    public void trySend(@NotNull Player player) {
        if (!spicordBotAvailable) return;
        DiscordBot bot = yolo.getSpicordManager().getSpicordBot();
        String embedFromTemplate;
        try {
            embedFromTemplate = deathMessageTemplate.replace("%player_name%", player.getName());
        } catch (NullPointerException npe){
            embedFromTemplate = Yolo.getPlugin(Yolo.class).getPluginResourceBundle().getString("sending.no_death_message");
        }
        net.dv8tion.jda.api.entities.MessageEmbed embed = EmbedParser.parse(embedFromTemplate).toJdaEmbed();
        if (embed.isSendable()) {
            try {
                MessageCreateAction messageCreateAction = Objects.requireNonNull(bot.getJda().getTextChannelById(messageChannelId)).sendMessage(MessageCreateData.fromEmbeds(embed));
                messageCreateAction.submit().whenComplete((message, throwable) -> {
                    // Handle potential errors
                    if (throwable != null) yolo.getLogger().severe(yolo.getPluginResourceBundle().getString("sending.failed").replace("%error%", throwable.toString()));
                });
            } catch (NullPointerException e) {
                yolo.getLogger().severe(yolo.getPluginResourceBundle().getString("sending.null_channel"));
            }
        }
    }

    /**
     * Package private method to register this plugin as an addon for Spicord and to set a few important fields for Spicord
     * support.
     */
    void loadSpicord() {
        spicordBotAvailable = false;

        if (!yolo.getConfig().getBoolean("spicord.send")) return;

        // Get and validate the channel-id:
        messageChannelId = yolo.getConfig().getString("spicord.message_channel_id");
        boolean validId = false;
        if (messageChannelId != null) {
            if (messageChannelId.matches("[0-9]+")) {
                validId = true;
            }
        }
        if (!validId) {
            yolo.getLogger().warning(yolo.getPluginResourceBundle().getString("loading.spicord.invalidId"));
            return;
        }

        // Load the message template:
        try {
            updateDeathMessageTemplate();
        } catch (IOException e) {
            // Shouldn't happen
            yolo.getLogger().severe(
                    yolo.getPluginResourceBundle()
                            .getString("loading.spicord.failedMessageTemplate")
                            .replace("%error%", e.toString())
            );
            e.printStackTrace();
            return;
        }

        // Up until this point, the field spicordBotAvailable should be false;
        // Provide the spicord loader an addon for use with this plugin.
        SpicordLoader.addStartupListener(spicord -> spicord.getAddonManager().registerAddon(new SimpleAddon("Yolo-Spicord", "yolo-deaths", "eingruenesbeb", "v0.5.0") {
            @Override
            public void onReady(DiscordBot bot) {
                spicordBot = bot;
                spicordBotAvailable = bot != null;
                if (!spicordBotAvailable) {
                    yolo.getLogger().warning(yolo.getPluginResourceBundle().getString("loading.spicord.bot_unavailable"));
                } else {
                    yolo.getLogger().info(yolo.getPluginResourceBundle().getString("loading.spicord.bot_available"));
                }
            }

            @Override
            public void onShutdown(DiscordBot bot) {
                spicordBotAvailable = false;
            }

            @Override
            public void onDisable() {
                spicordBotAvailable = false;
            }
        }));
    }
}

File: src/main/kotlin/io/github/eingruenesbeb/yolo/managers/spicord/SpicordManager.kt
/*
 * This program is a plugin for Minecraft Servers called "Yolo".
 * Copyright (C) 2023-2023  eingruenesbeb
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 *
 * You can reach the original author via e-Mail: agreenbeb@gmail.com
 */
package io.github.eingruenesbeb.yolo.managers.spicord

import io.github.eingruenesbeb.yolo.Yolo
import io.github.eingruenesbeb.yolo.managers.ReloadableManager
import net.dv8tion.jda.api.entities.Message
import net.dv8tion.jda.api.entities.channel.concrete.TextChannel
import net.dv8tion.jda.api.utils.messages.MessageCreateData
import org.bukkit.plugin.java.JavaPlugin
import org.spicord.Spicord
import org.spicord.SpicordLoader
import org.spicord.api.addon.SimpleAddon
import org.spicord.bot.DiscordBot
import org.spicord.embed.Embed
import org.spicord.embed.EmbedParser
import java.io.IOException
import java.nio.file.Files
import java.util.*

/**
 * This object manages every aspect regarding functionality for sending Discord messages via Spicord.
 */
internal object SpicordManager : ReloadableManager {
    /**
     * Used to store the raw string data and enabled status of a discord embed. All potential [RawDiscordEmbed]s are
     * stored within the private value [rawDiscordEmbedEnumMap] for look-ups and don't have any other apparent use
     * cases.
     */
    @JvmRecord
    data class RawDiscordEmbed(val rawString: String, val enabled: Boolean) {
        /**
         * Converts the [rawString] into an [Embed] from Spicord, whilst applying potential replacements, which may be
         * provided by [io.github.eingruenesbeb.yolo.TextReplacements.provideStringDefaults]-.
         */
        fun returnSpicordEmbed(replacements: HashMap<String, String?>?): Embed {
            var toParse = rawString
            replacements?.keys?.forEach { replacementKey ->
                replacements[replacementKey]?.let { toParse = toParse.replace(replacementKey, it) }
            }
            return EmbedParser.parse(toParse)
        }
    }

    /**
     * If the bot is available for sending a message.
     */
    var isSpicordBotAvailable = false
        private set

    /**
     * (Spicord version of) The bot, that is used for sending the message, if provided.
     *
     * @see DiscordBot
     */
    var spicordBot: DiscordBot? = null
        private set

    private val yolo = JavaPlugin.getPlugin(Yolo::class.java)
    private val rawDiscordEmbedEnumMap = EnumMap<DiscordMessageType, RawDiscordEmbed>(DiscordMessageType::class.java)
    private var messageChannelId: String? = null

    init {
        loadSpicord()
    }

    /**
     * Tries to send a message to Discord created from the configured template, using Spicord and JDA.
     *
     * @param discordMessageType The Type of the message to be sent, as defined by [DiscordMessageType].
     * @param replacements The replacements to perform. (Placeholders typically look like "%example%".)
     */
    fun trySend(discordMessageType: DiscordMessageType, replacements: HashMap<String, String?>?) {
        if (!isSpicordBotAvailable) return
        if (!yolo.config.getBoolean("spicord.send")) return
        val rawDiscordEmbed = rawDiscordEmbedEnumMap[discordMessageType]
        if (!rawDiscordEmbed!!.enabled) return
        val toSend = rawDiscordEmbed.returnSpicordEmbed(replacements).toJdaEmbed()
        if (toSend.isSendable) {
            try {
                val textChannel = spicordBot!!.jda.getGuildChannelById(messageChannelId!!) as? TextChannel
                val messageCreateAction = textChannel!!.sendMessage(MessageCreateData.fromEmbeds(toSend))
                messageCreateAction.submit().whenComplete { _: Message?, throwable: Throwable? ->
                    // Handle potential errors
                    if (throwable != null) yolo.logger.severe {
                        Yolo.pluginResourceBundle.getString("spicord.sending.failed")
                            .replace("%error%", throwable.toString())
                    }
                }
            } catch (e: NullPointerException) {
                yolo.logger.severe {Yolo.pluginResourceBundle.getString("spicord.sending.nullChannel") }
            }
        }
    }

    override fun reload() {
        isSpicordBotAvailable = false

        // Get and validate the channel-id:
        messageChannelId = yolo.config.getString("spicord.message_channel_id")
        if (messageChannelId?.matches("[0-9]+".toRegex()) != true) {
            yolo.logger.warning { Yolo.pluginResourceBundle.getString("loading.spicord.invalidId") }
            return
        }

        // Load the message templates:
        try {
            updateMessageTemplates()
        } catch (e: IOException) {
            // Shouldn't happen
            yolo.logger.severe {
                Yolo.pluginResourceBundle
                    .getString("loading.spicord.failedMessageTemplate")
                    .replace("%error%", e.toString())
            }
            e.printStackTrace()
            return
        }
        isSpicordBotAvailable = spicordBot!!.isReady
    }

    @Throws(IOException::class)
    private fun updateMessageTemplates() {
        rawDiscordEmbedEnumMap.clear()
        for (discordMessageType in DiscordMessageType.entries) {
            val resourceName: String = discordMessageType.resourceName
            val resourcePath = yolo.dataFolder.toPath().resolve(resourceName)
            val contents: String = try {
                Files.readString(resourcePath)
            } catch (e: IOException) {
                // Default to the embedded resource.
                // Embedded resources should not throw an NPE.
                String(yolo.getResource(resourceName)!!.readAllBytes())
            }
            val rawDiscordEmbed = RawDiscordEmbed(contents, yolo.config.getBoolean(discordMessageType.enabledKey))
            rawDiscordEmbedEnumMap[discordMessageType] = rawDiscordEmbed
        }
    }

    /**
     * Private method to register this plugin as an addon for Spicord and to set a few important fields for Spicord
     * support.
     */
    private fun loadSpicord() {
        isSpicordBotAvailable = false

        // Get and validate the channel-id:
        messageChannelId = yolo.config.getString("spicord.message_channel_id")
        if (messageChannelId?.matches("[0-9]+".toRegex()) != true) {
            yolo.logger.warning { Yolo.pluginResourceBundle.getString("loading.spicord.invalidId") }
            return
        }

        // Load the message templates:
        try {
            updateMessageTemplates()
        } catch (e: IOException) {
            // Shouldn't happen
            yolo.logger.severe {
                Yolo.pluginResourceBundle
                    .getString("loading.spicord.failedMessageTemplate")
                    .replace("%error%", e.toString())
            }
            e.printStackTrace()
            return
        }

        // Up until this point, the field spicordBotAvailable should be false;
        // Provide the spicord loader an addon for use with this plugin.
        SpicordLoader.addStartupListener { spicord: Spicord ->
            spicord.addonManager.registerAddon(object : SimpleAddon("Yolo-Spicord", "yolo", "eingruenesbeb", "v0.5.0") {
                override fun onReady(bot: DiscordBot) {
                    spicordBot = bot
                    isSpicordBotAvailable = true
                    if (!isSpicordBotAvailable) {
                        yolo.logger.warning { Yolo.pluginResourceBundle.getString("loading.spicord.bot_unavailable") }
                    } else if (yolo.config.getBoolean("spicord.send")) {
                        yolo.logger.info { Yolo.pluginResourceBundle.getString("loading.spicord.bot_available") }
                    }
                }

                override fun onShutdown(bot: DiscordBot) {
                    isSpicordBotAvailable = false
                }

                override fun onDisable() {
                    isSpicordBotAvailable = false
                }
            })
        }
    }
}
